### DISCUSSION TITLE
Need help to decode data

### DISCUSSION THREAD

**Discussion by dpoitou-tlse** [jasonacox/tinytuya#338] [Q&A] [answered]
Hello,
I'm trying to use tinytuya to get data from a flood detector
https://www.amazon.fr/Alupre-D%C3%A9tecteur-d%C3%A9bordement-Inondation-Accueil/dp/B08HS19K7W/ref=sr_1_13?__mk_fr_FR=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=RQSXNEJ2ZGNT&keywords=detecteur+fuite+tuya&qid=1682429861&sprefix=detecteur+fuite+tuya%2Caps%2C124&sr=8-13

I'm using it to check the level of my pool. To have a better detection I replaced the sensor by a switch like that
https://www.amazon.fr/Broco-Interrupteurs-interrupteur-horizontal-r%C3%A9servoir/dp/B082917X8D/ref=sr_1_4?__mk_fr_FR=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=YOSX2AM6F9QT&keywords=switch+eau+flotteur&qid=1682429987&sprefix=switch+eau+flotteur%2Caps%2C92&sr=8-4

I placed the wifi part of the sensor in a sealed box of my skimmer. Now I'm trying to get the information of the sensor.- In the app it is working well.
I tried to use tinytuya in local mode but as the sensor is working with a battery it is off most of the time and I did not succeded to get data directly.

So I tried to use the cloud mode and I think it is working well

```
c = tinytuya.Cloud()  # Connect to Tuya Cloud, uses tinytuya.json 

devices = c.getdevices() # Display list of devices
print("Device List: %r" % devices)

id = "xxxxxxxxxxxxxxxxxxID" # Select a Device ID to Test

result = c.getstatus(id) # Display Status of Device
print("Status of device:\n", result)
```

I get
`Status of device:
 {'result': [{'code': 'control', 'value': 'stop'}], 'success': True, 't': 1682430287920, 'tid': '582edfc9e36f11ed8e5bfedce89f50b1'}`

Could you help me to decode t or tid to get the state of the sensor please ?

Regards

**Comment by uzlonewolf** [comment]
`t` is just the transaction Unix timestamp * 1000, so 1682430287920 = 1682430287.920 = 2023-04-25 06:44:47
`tid` is just a unique identifier for that transaction so you can pull the error logs if needed.

The data you want is going to be in that `result` key.  It looks like it has a single DP, called `control`, and currently has the value `stop`.  You can get more info on it by calling `.getproperties(deviceid)` and/or `.getdps(deviceid)`.  Historic data can also be retrieved with `.getdevicelog(deviceid)`.

If the device has additional DPs which are not showing up via the cloud, you can change "Control Instruction Mode" from "Standard Instruction" to "DP Instruction" to get them  https://github.com/jasonacox/tinytuya/discussions/284#discussioncomment-4953888

**Comment by dpoitou-tlse** [comment]
I tried the different functions to get the state of the sensor
```
Status of device c.getdevicelog:
 {'result': {'device_id': '580422722462ab0510b7', 'has_next': False, 'logs': []}, 'success': True, 't': 1683050041457, 'tid': '520ec688e91211edacb1ee0c9cb2bdce', 'fetches': 1} 

Status of device c.getproperties:
 {'result': {'category': 'cl', 'functions': [{'code': 'control', 'desc': '{"range":["open","stop","close"]}', 'name': '控制', 'type': 'Enum', 'values': '{"range":["open","stop","close"]}'}], 'status': [{'code': 'control', 'name': '控制', 'type': 'Enum', 'values': '{"range":["open","stop","close"]}'}]}, 'success': True, 't': 1683050041793, 'tid': '524b5bcce91211ed9b43a28304c1a187'} 

Status of device c.getstatus:
 {'result': [{'code': 'control', 'value': 'stop'}], 'success': True, 't': 1683050042179, 'tid': '5284d928e91211ed9b43a28304c1a187'} 

Status of device c.getdps:
 {'result': {'category': 'cl', 'functions': [{'code': 'control', 'dp_id': 1, 'type': 'Enum', 'values': '{"range":["open","stop","close"]}'}], 'status': [{'code': 'control', 'dp_id': 1, 'type': 'Enum', 'values': '{"range":["open","stop","close"]}'}]}, 'success': True, 't': 1683050042603, 'tid': '52c6164fe91211ed9b43a28304c1a187'} 

Status of device c.getconnectstatus:
 True 
```

As in the post #282 I tried to switch to DP mode under the Tuya IOT plateform:
![Capture d’écran 2023-05-02 à 19 57 27](https://user-images.githubusercontent.com/55097404/235748491-e1b8c6fa-430b-4b07-94dc-d79e3062c4c7.png)
Does not change the return of the above functions but I did not waited 12-24h after the change... However the information seems to be the same between normal and DP modes.

I can see the state of the sensor under Tuya IOT plateform and correclty updated when changing the sensor physically going into Device debugging / Update Device Status : 
![Capture d’écran 2023-05-02 à 20 01 18](https://user-images.githubusercontent.com/55097404/235749003-fc8da59a-e456-4e95-94c1-2ebc93ac4e6c.png)

In the readme, section "Tuya Data Points - DPS Table" there is this table speaking about flooding sensor:
![Capture d’écran 2023-05-02 à 20 09 11](https://user-images.githubusercontent.com/55097404/235749619-232fedf2-419f-4144-9ccf-b36b9ddab733.png)

But I do not understand how to use the "DP ID" in cloud mode. As my sensor works on battery I think the cloud mode is the only was to get informations from it (ie. ping is not responding on the sensor, only when the state changes).

**Reply by dpoitou-tlse** [reply]
A view of my pool level sensor adapted from the flooding sensor:
![IMG_5486](https://github.com/jasonacox/tinytuya/assets/55097404/f8b16281-87c2-4a87-a5e6-4876ce09cc61)
![IMG_5487](https://github.com/jasonacox/tinytuya/assets/55097404/5b53e108-9159-427f-89a8-4b1e14c88e9b)
 
Is there a way to retrieve a json data directly from the website?

Thanks for your help

**Reply by uzlonewolf** [reply]
Like most modules, it would seem leak detectors come in several different styles.  Unfortunately that particular one is a different type than the one listed in the DPS table.  That `"range":["open","stop","close"]` line makes me think it is designed to operate a shut-off valve; when it detects water it goes into `close` mode for a few seconds before going back into `stop` mode once the valve closes.  Once the water goes away it will probably go `open` before going `stop` again.

Unfortunately, without having one to play with myself, I don't think I can offer much help.  Does the "Device Logs" tab on Tuya's site show any of the activations?  If so then you should be able to tweak the `.getdevicelog()` parameters to fetch them that way.

**Reply by dpoitou-tlse** [reply]
I made some progress...

On tuya's site I got the state of my sensor

[Capture d’écran 2023-05-16 à 23.46.41.pdf](https://github.com/jasonacox/tinytuya/files/11492321/Capture.d.ecran.2023-05-16.a.23.46.41.pdf)

Translating the curl command in python gives me:
```
import requests
import json
import time
 
# ts stores the time in seconds
ts = int(time.time()*1000)

headers = {
    'sign_method': 'HMAC-SHA256',
    'client_id': 'XXXX',
    't': '1684273879885',
    'mode': 'cors',
    'Content-Type': 'application/json',
    'sign': '9755FB7AFB6DFE296C498EDA279FC09B5063B398D7C069E6280DD8FF022789BF',
    'access_token': 'XXXX',
}
response = requests.get('https://openapi.tuyaeu.com/v1.0/iot-03/devices/82548706ecfabcbce14b/status', headers=headers)

print(response.text)

data = json.loads(response.text)
valeur_alarme = data["result"][0]["value"]

print(valeur_alarme)
```
The result is 
```
{"result":[{"code":"watersensor_state","value":"alarm"},{"code":"battery_state","value":"high"}],"success":true,"t":1684273921371,"tid":"e34fb463f43311ed9b213e97a7ff9260"}
alarm
```
the watersensor_state is in data["result"][0]["value"] as I was looking for 👍 

My problem is that the sign value is generated by the IOT plateform linked to the timestamp.

Would you know how to generate the "sign" value properly?

**Reply by dpoitou-tlse** [reply]
I found this page
https://developer.tuya.com/en/docs/iot/new-singnature?id=Kbw0q34cs2e5g

But I think the tinytuya should do something similar...

I try to understand how to generate the signature..

**Reply by dpoitou-tlse** [reply]
Tried something like that
```
import hashlib
import hmac
import time
 
# ts stores the time in seconds
t = int(time.time()*1000)
print(t)
client_id='XXX'
access_token='XXX'

concatenated_string = client_id + access_token + str(t)

message = "concatenated_string"
secret = 'XXX'

# Calculer le HMAC-SHA256
hmac_sha256 = hmac.new(bytes(secret, 'utf-8'), bytes(message, 'utf-8'), hashlib.sha256)
s = hmac_sha256.hexdigest().upper()

print(s)
```

Not working... Tired..

**Reply by uzlonewolf** [reply]
I'm not sure what it is you are trying to do?  Calling `c.getstatus(id)` gets that exact same URL and does the `sign` calculation for you, and even if you insist on sending it yourself you can just fetch it with `c.cloudrequest( '/v1.0/iot-03/devices/82548706ecfabcbce14b/status' )`.

Have you tried `c.getstatus(id)` again now that it has been more than 24 hours since changing the Control Instruction Mode?

**Reply by dpoitou-tlse** [reply, marked as answer]
Ok it is working now..
I think there was a mistake on the id of my sensor.. I don't understand why but it is working now !
